text/0000-return-type-notation.md

```diff
@@ -0,0 +1,1105 @@
+# Return Type Notation in Bounds and Where-clauses
+
+- Feature Name: `return_type_notation`
+- Start Date: 2024-06-04
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Return type notation (RTN) gives a way to reference or bound the type returned by a trait method.
+The new bounds look like `T: Trait<method(..): Send>` or `T::method(..): Send`.
```

Comment:
I think it could be cleaner to leverage [`FnOnce::Output`](https://doc.rust-lang.org/std/ops/trait.FnOnce.html#associatedtype.Output) somehow, which then wouldn't require any new syntax. E.g. `T::method::Output: Send` desguaring to something like `<<T as Trait>::method as FnOnce(..) -> _>::Output: Send`. 

Unfortunately similar patterns hit "the precise format of `Fn`-family traits' type parameters is subject to change". Not sure how feasible it would be to consider at least the `Output` associated type stable, but I would _love_ to be able to use this elsewhere too.

Reply:
We explored this option. This exploration is covered under the FAQ ["What other syntax options were considered"](https://github.com/nikomatsakis/rfcs/blob/rtn/text/0000-return-type-notation.md#what-other-syntax-options-were-considered) which, for this particular proposal, bottomed out in a [design like this one](https://smallcultfollowing.com/babysteps/blog/2023/06/12/higher-ranked-projections-send-bound-problem-part-4/). One major drawback noted in the RFC:

> This reuses associated types but, as both the function and future traits define an `Output` associated type, raises the potential for confusion about whether this notation means "the future that gets returned" or "the result of the future".

We ultimately decided against this -- but we also noted that it could be added in the future, in which case `(..)` could be seen as a shorthand for `::Output`, essentially.